File: cpp/linked-list/easy/234.回文链表.cpp
/*
 * @lc app=leetcode.cn id=234 lang=cpp
 *
 * [234] 回文链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *left = nullptr;

    bool isPalindrome(ListNode *head)
    {
        left = head;
        return traverse(head);
    }

    bool traverse(ListNode *right)
    {
        if (right == nullptr) {
            return true;
        }
        // 后序递归遍历
        auto res = traverse(right->next);
        res      = res && (left->val == right->val);
        left     = left->next;
        return res;
    }
};
// @lc code=end

File: golang/linked-list/easy/234.回文链表.go
/*
 * @lc app=leetcode.cn id=234 lang=golang
 * @lcpr version=30202
 *
 * [234] 回文链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	left := head
	var right *ListNode

	res := true
	var traverse func(*ListNode)
	traverse = func(r *ListNode) {
		if r == nil {
			return
		}

		traverse(r.Next)

		if left.Val != r.Val {
			res = false
			return
		}

		left = left.Next
	}

	right = head
	traverse(right)
	return res
}

// @lc code=end

/*
// @lcpr case=start
// [1,2,2,1]\n
// @lcpr case=end

// @lcpr case=start
// [1,2]\n
// @lcpr case=end

*/

